src/Storages/Kafka/StorageKafka2.cpp

```diff
@@ -745,63 +746,204 @@ void StorageKafka2::dropReplica()
+    }
+
+    return locked_partitions;
+}
+
 
-        // It is okay that these paths are created in a different transaction. The important thing is the lock file.
+void StorageKafka2::createLocksInfo(zkutil::ZooKeeper & keeper_to_use, TopicPartitionLocks & locks, const TopicPartition& partition_to_lock, bool new_lock)
+{
+    const auto topic_partition_path = getTopicPartitionPath(partition_to_lock);
+    if (new_lock) {
+        const auto lock_file_path = String(topic_partition_path / lock_file_name);
         keeper_to_use.createAncestors(lock_file_path);
+        const auto lock_replica_name = (*kafka_settings)[KafkaSetting::kafka_replica_name].value;
+        Coordination::Error code = keeper_to_use.tryCreate(lock_file_path, lock_replica_name, zkutil::CreateMode::Ephemeral);
+        if (code != Coordination::Error::ZOK)
+        {
+            if (code != Coordination::Error::ZNODEEXISTS)
+                throw zkutil::KeeperException(code);
 
-        ops.push_back(zkutil::makeCreateRequest(lock_file_path, (*kafka_settings)[KafkaSetting::kafka_replica_name].value, zkutil::CreateMode::Ephemeral));
-        LOG_TRACE(log, "Creating locking ops for: {}", lock_file_path);
+            LOG_TRACE(
+                log,
+                "Couldn't create lock for topic partition {}:{} from replica {} because it already exists",
+                partition_to_lock.topic,
+                partition_to_lock.partition_id,
+                lock_replica_name
+            );
+            return;
+        }
     }
 
-    Coordination::Responses responses;
+    using zkutil::EphemeralNodeHolder;
+    LockedTopicPartitionInfo lock_info{
+        EphemeralNodeHolder::existing(topic_partition_path / lock_file_name, keeper_to_use),
+        getNumber(keeper_to_use, topic_partition_path / commit_file_name),
+        getNumber(keeper_to_use, topic_partition_path / intent_file_name)};
+
+    LOG_TRACE(
+        log,
+        "Locked topic partition: {}:{} at offset {} with intent size {}",
+        partition_to_lock.topic,
+        partition_to_lock.partition_id,
+        lock_info.committed_offset.value_or(0),
+        lock_info.intent_size.value_or(0));
+    locks.emplace(TopicPartition(partition_to_lock), std::move(lock_info));
+}
+
+
+void StorageKafka2::unlockTopicPartition(zkutil::ZooKeeper & keeper_to_use,  const TopicPartition& partition_to_unlock) {
+    const auto topic_partition_path = getTopicPartitionPath(partition_to_unlock);
+    const auto lock_file_path = String(topic_partition_path / lock_file_name);
+    Coordination::Error code = keeper_to_use.tryRemove(lock_file_path);
+    if (code == Coordination::Error::ZNONODE)
+        return;
+
+    LOG_TRACE(
+        log,
+        "Replica {} successfully unlocked topic partition {}:{}",
+        (*kafka_settings)[KafkaSetting::kafka_replica_name].value,
+        partition_to_unlock.topic,
+        partition_to_unlock.partition_id
+    );
+}
```

Comment:
The ephemeral nodes are removed when the keeper session expires or when the `EphemeralNodeHolder` is destructed. Removing "lock files" manually is a sign that something is not going good.
```suggestion
```